File: android/app/src/main/kotlin/com/debrify/app/tv/ShowShuffle.kt
package com.debrify.app.tv

import kotlin.random.Random

internal data class ShuffleEpisode(val season: Int, val episode: Int)

/** Episode identities survive replacement of the current torrent's file list. */
internal class ShowShuffle(private val random: Random = Random.Default) {
    private val visited = mutableSetOf<ShuffleEpisode>()

    fun clear() = visited.clear()

    fun pick(
        episodes: List<ShuffleEpisode>,
        current: ShuffleEpisode?,
        excluded: Set<ShuffleEpisode> = emptySet(),
    ): ShuffleEpisode? {
        val eligible = episodes.toSet() - excluded - setOfNotNull(current)
        if (eligible.isEmpty()) return null
        current?.let { visited.add(it) }
        var remaining = eligible - visited
        if (remaining.isEmpty()) {
            visited.clear()
            current?.let { visited.add(it) }
            remaining = eligible
        }
        return remaining.toList().random(random).also { visited.add(it) }
    }
}

File: lib/utils/show_shuffle.dart
import 'dart:math';

/// Stable identities keep shuffle history independent of the currently open pack.
typedef ShuffleEpisode = (int, int);

bool isShuffleEpisodeEligible(Map<String, dynamic> episode, {DateTime? now}) {
  final season = episode['season'];
  final number = episode['number'];
  if (season is! int || number is! int || season <= 0 || number <= 0) {
    return false;
  }
  final air = DateTime.tryParse(
    (episode['airstamp'] ?? episode['airdate'] ?? '').toString(),
  );
  // Unknown dates remain playable; positively future-dated episodes do not.
  return air == null || !air.isAfter(now ?? DateTime.now());
}

class ShowShuffle {
  ShowShuffle({Random? random}) : _random = random ?? Random();
  final Random _random;
  final Set<ShuffleEpisode> _visited = {};

  void clear() => _visited.clear();

  ShuffleEpisode? pick(
    Iterable<ShuffleEpisode> episodes,
    ShuffleEpisode? current, {
    Set<ShuffleEpisode> excluded = const {},
  }) {
    final eligible = episodes.toSet()..removeAll(excluded);
    eligible.remove(current);
    if (eligible.isEmpty) return null;
    if (current != null) _visited.add(current);
    var remaining = eligible.difference(_visited).toList();
    if (remaining.isEmpty) {
      _visited.clear();
      if (current != null) _visited.add(current);
      remaining = eligible.toList();
    }
    final next = remaining[_random.nextInt(remaining.length)];
    _visited.add(next);
    return next;
  }
}
